pkg/util/system/system_utils.go

```diff
@@ -17,21 +17,16 @@ limitations under the License.
 package system
 
 import (
-	"strings"
+	"k8s.io/api/core/v1"
 )
 
+const LabelNodeRoleMaster = "node-role.kubernetes.io/master"
+
 // IsMasterNode returns true if given node is a registered master.
-// TODO: find a better way of figuring out if given node is a registered master.
-func IsMasterNode(nodeName string) bool {
-	// We are trying to capture "master(-...)?$" regexp.
-	// However, using regexp.MatchString() results even in more than 35%
-	// of all space allocations in ControllerManager spent in this function.
-	// That's why we are trying to be a bit smarter.
-	if strings.HasSuffix(nodeName, "master") {
+// use node-role.kubernetes.io/master label to figure out if given node is a registered master.
+func IsMasterNode(node *v1.Node) bool {
+	if _, ok := node.Labels[LabelNodeRoleMaster]; ok {
 		return true
 	}
-	if len(nodeName) >= 10 {
```

Comment:
If e2e tests are depending on this we should probably also preserve the previous detection method...

Reply:
So in this PR, you mean that just  write another new IsMasterNode func, and make the pkg use the new IsMasterNode func. And the e2e tests preserve use the previous IsMasterNode for compatibility (I saw the perf-test repo also depend on this func). Then clean the old func after removing the dependencies.

Reply:
er no, I mean we can perform the OR of both checks